source/common/upstream/upstream_impl.cc

```diff
@@ -623,20 +624,40 @@ void HostSetImpl::rebuildLocalityScheduler(
   locality_scheduler = nullptr;
   if (all_hosts_per_locality != nullptr && locality_weights != nullptr &&
       !locality_weights->empty() && !eligible_hosts.empty()) {
-    locality_scheduler = std::make_unique<EdfScheduler<LocalityEntry>>();
-    locality_entries.clear();
-    for (uint32_t i = 0; i < all_hosts_per_locality->get().size(); ++i) {
-      const double effective_weight = effectiveLocalityWeight(
-          i, eligible_hosts_per_locality, *excluded_hosts_per_locality, *all_hosts_per_locality,
-          *locality_weights, overprovisioning_factor);
-      if (effective_weight > 0) {
-        locality_entries.emplace_back(std::make_shared<LocalityEntry>(i, effective_weight));
-        locality_scheduler->add(effective_weight, locality_entries.back());
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.edf_lb_locality_scheduler_init_fix")) {
+      locality_entries.clear();
+      for (uint32_t i = 0; i < all_hosts_per_locality->get().size(); ++i) {
+        const double effective_weight = effectiveLocalityWeight(
+            i, eligible_hosts_per_locality, *excluded_hosts_per_locality, *all_hosts_per_locality,
+            *locality_weights, overprovisioning_factor);
+        if (effective_weight > 0) {
+          locality_entries.emplace_back(std::make_shared<LocalityEntry>(i, effective_weight));
+        }
+      }
+      // If not all effective weights were zero, create the scheduler.
+      if (!locality_entries.empty()) {
+        locality_scheduler = std::make_unique<EdfScheduler<LocalityEntry>>(
+            EdfScheduler<LocalityEntry>::createWithPicks(
+                locality_entries,
+                [](const LocalityEntry& entry) { return entry.effective_weight_; }, seed));
```

Comment:
do we need to ensure that `seed` > 0?

Reply:
`seed` is defined as `uint64_t` in the input of the current function, and as a `uint32_t` in the input of `createWithPick()`, so both are unsigned and should be >0. The implicit casting seems fine to me, as it just specifies the number of initial-picks which if it is more than 10^6 is probably bad anyway.

Reply:
This was validated in [this test](https://github.com/envoyproxy/envoy/blob/924f778c9ad8c5a4d8d1c6af2a54012b238a9b14/test/common/upstream/edf_scheduler_test.cc#L180).